### PR TITLE
fix #1518 Add cold variants of TestPublisher

### LINKED ISSUE #1518: Add TestPublisher.createCold variants that backpressure or overflow
`TestPublisher.createCold()` currently creates a `ColdTestPublisher` that ignores requests and replays the whole buffer to new incoming subscribers, regardless of `request()`.

The base version could create a copy of the buffer and use it to queue elements so as to respect the request (onBackpressureBuffer, basically).

Additionally, it could be parameterized to misbehave (like it currently does) or to fail on overflow (onBackpressureError).

The 3 variants would be exposed as factory methods on `TestPublisher`, changing the default behavior to that of a backpressure-enforcing cold Publisher, much like `Flux.from(...)`.

### PR DESCRIPTION
**Pull request by ericbottard**

### CODE REVIEW

File: reactor-test/src/main/java/reactor/test/publisher/ColdTestPublisher.java [hunk 1]
```diff
@@ -226,6 +221,76 @@ void onComplete() {
...
+					if (cancelled) {
+						return;
+					}
+				}
+
+				n = requested;
+				if (emitted > n) {
+					assert parent.violations.contains(Violation.REQUEST_OVERFLOW);
+					parent.hasOverflown = true;
+					return;
```
Comment (simonbasle) [inline reactor-test/src/main/java/reactor/test/publisher/ColdTestPublisher.java:256]: shouldn't it check for `emitTerminalSignalIfAny()` here too?

File: reactor-test/src/main/java/reactor/test/publisher/ColdTestPublisher.java [hunk 2]
```diff
@@ -338,23 +403,29 @@ public boolean wasRequested() {
...
 	public ColdTestPublisher<T> next(@Nullable T t) {
-		Objects.requireNonNull(t, "emitted values must be non-null");
+		if (!violations.contains(ALLOW_NULL)) {
+			Objects.requireNonNull(t, "emitted values must be non-null");
+		}
 
 		values.add(t);
 		for (ColdTestPublisherSubscription<T> s : subscribers) {
-			s.onNext(t);
+			s.drain(s.requested);
```
Comment (simonbasle) [inline reactor-test/src/main/java/reactor/test/publisher/ColdTestPublisher.java:428]: wouldn't it be possible for the `drain` to have no parameter, computing the `n` from its latest `requested` state?

File: reactor-test/src/main/java/reactor/test/publisher/ColdTestPublisher.java [hunk 3]
```diff
@@ -226,6 +221,76 @@ void onComplete() {
 			parent.remove(this);
 			actual.onComplete();
 		}
+
+		private void drain(long n) {
```
Comment (simonbasle) [inline reactor-test/src/main/java/reactor/test/publisher/ColdTestPublisher.java:225]: drain has to be guarded by a WIP in order to avoid parallel drains (since there is no restriction on which thread calls `next` from the `TestPublisher` API vs `request` from the `Subscription` API)

File: reactor-test/src/main/java/reactor/test/publisher/TestPublisher.java [hunk 4]
```diff
@@ -72,13 +75,43 @@
...
+	 * @param <T> the type of the publisher
+	 * @return the new {@link TestPublisher}
+	 */
+	public static <T> TestPublisher<T> createColdNonBuffering() {
+		return new ColdTestPublisher<>(true, EnumSet.noneOf(Violation.class));
+	}
+
+	/**
+	 * Create a cold {@link TestPublisher}, which can be subscribed to by multiple
+	 * subscribers. It caches the {@link #next(Object)} events and replays them to
```
Comment (simonbasle) [inline reactor-test/src/main/java/reactor/test/publisher/TestPublisher.java:103]: it only caches if `boolean errorOnOverflow` is `false`

File: reactor-test/src/test/java/reactor/test/publisher/ColdTestPublisherTests.java [hunk 5]
```diff
@@ -111,6 +124,21 @@ public void coldDisallowsOverflow() {
 		publisher.assertNoRequestOverflow();
 	}
 
+	@Test
+	public void misbehavingAllowsOverflow() {
+		TestPublisher<String> publisher = TestPublisher.createColdNonCompliant(false, TestPublisher.Violation.REQUEST_OVERFLOW);
+
+		assertThatExceptionOfType(AssertionError.class)
```
Comment (simonbasle) [inline reactor-test/src/test/java/reactor/test/publisher/ColdTestPublisherTests.java:131]: If that test is made that way because the `StepVerifier` complains that its under-requesting when adding an `expectNext("bar")`, that under-requesting check can be disabled using `StepVerifierOptions` (namely, `StepVerifierOptions#checkUnderRequesting(false)`).

**Review by simonbasle — CHANGES_REQUESTED** [review]

File: reactor-test/src/main/java/reactor/test/publisher/ColdTestPublisher.java (hunk 1 above)
Comment (ericbottard) [reply reactor-test/src/main/java/reactor/test/publisher/ColdTestPublisher.java:256]: It could indeed. That branch is the very exceptional case I thought, leaving the publisher in a "bad state" state, but it could go the extra mile and also emit terminal signals (which should be also protected against)

File: reactor-test/src/main/java/reactor/test/publisher/TestPublisher.java (hunk 4 above)
Comment (ericbottard) [reply reactor-test/src/main/java/reactor/test/publisher/TestPublisher.java:103]: No, it accumulates values and replays them to new subscribers in all cases. The boolean flag only drives what happens when there is more (total) request than values in the `values` buffer: wait for more request, or `error()`.

File: reactor-test/src/test/java/reactor/test/publisher/ColdTestPublisherTests.java (hunk 5 above)
Comment (ericbottard) [reply reactor-test/src/test/java/reactor/test/publisher/ColdTestPublisherTests.java:131]: I'm not sure I get what you mean by "made that way" (do you mean using `assertThatExceptionOfType().isThrownBy()` (ie direct exception)?).
That test was copied and adapted from the `DefaultTestPublisher` IIRC.

File: reactor-test/src/main/java/reactor/test/publisher/ColdTestPublisher.java (hunk 2 above)
Comment (ericbottard) [reply reactor-test/src/main/java/reactor/test/publisher/ColdTestPublisher.java:428]: I've always been puzzled by that myself everywhere we see this construct, but I guess for the case where it is called from `request(n)` we can't, can we?
```java
		public void request(long n) {
			if (Operators.validate(n)) {
				if (Operators.addCap(REQUESTED, this, n) == 0) {
					parent.wasRequested = true;
					drain(n);
				}
			}
		}
```

Comment (ericbottard) [reply reactor-test/src/main/java/reactor/test/publisher/ColdTestPublisher.java:428]: Actually, we can, since it is only called if total request *now* equals `n`. Hmmm 🤔

File: reactor-test/src/main/java/reactor/test/publisher/TestPublisher.java (hunk 4 above)
Comment (simonbasle) [reply reactor-test/src/main/java/reactor/test/publisher/TestPublisher.java:103]: ok then I'd rephrase the second paragraph in `createCold` and `createColdNonBuffering` above, then. Will comment there.

File: reactor-test/src/main/java/reactor/test/publisher/TestPublisher.java [hunk 6]
```diff
@@ -72,13 +75,43 @@
 	 * subscribers. It caches the {@link #next(Object)} events and replays them to
```
Comment (simonbasle) [inline reactor-test/src/main/java/reactor/test/publisher/TestPublisher.java:75]: Let's use `buffers` instead of `caches`. Also, rephrase the `and replays...` part to something more along the lines of:
> and tracks how many elements have been seen by each subscriber in order to correctly replay the buffer to each subscriber, with backpressure.

File: reactor-test/src/main/java/reactor/test/publisher/TestPublisher.java [hunk 7]
```diff
@@ -72,13 +75,43 @@
...
 	 * all subscribers upon subscription.
 	 * <p>
-	 * Note that this type of {@link Publisher} isn't
+	 * The returned publisher will buffer values if there is not enough request from subscribers.
```
Comment (simonbasle) [inline reactor-test/src/main/java/reactor/test/publisher/TestPublisher.java:78]:
```suggestion
	 * The publisher honors backpressure, holding off emitting newest items
	 * from the buffer if the subscriber doesn't have enough request.
```

File: reactor-test/src/main/java/reactor/test/publisher/TestPublisher.java [hunk 8]
```diff
@@ -72,13 +75,43 @@
...
-		return new ColdTestPublisher<>();
+		return new ColdTestPublisher<>(false, EnumSet.noneOf(Violation.class));
+	}
+
+	/**
+	 * Create a cold {@link TestPublisher}, which can be subscribed to by multiple
+	 * subscribers. It caches the {@link #next(Object)} events and replays them to
+	 * all subscribers upon subscription.
+	 * <p>
+	 * The returned publisher will emit an overflow error if there is not enough request from subscribers.
```
Comment (simonbasle) [inline reactor-test/src/main/java/reactor/test/publisher/TestPublisher.java:92]:
```suggestion
	 * subscribers. It buffers the {@link #next(Object)} events and tracks how many
	 * elements have been seen by each subscriber in order to correctly replay the
	 * buffer.
	 * <p>
	 * The returned publisher will emit an overflow error if a new subscriber's
	 * first request is lower than the current buffer size, or if a new element
	 * is pushed to a registered subscriber that has zero pending demand.
```

File: reactor-test/src/main/java/reactor/test/publisher/TestPublisher.java [hunk 9]
```diff
@@ -72,13 +75,43 @@
...
+	 * @return the new {@link TestPublisher}
+	 */
+	public static <T> TestPublisher<T> createColdNonBuffering() {
+		return new ColdTestPublisher<>(true, EnumSet.noneOf(Violation.class));
+	}
+
+	/**
+	 * Create a cold {@link TestPublisher}, which can be subscribed to by multiple
+	 * subscribers. It caches the {@link #next(Object)} events and replays them to
+	 * all subscribers upon subscription.
```
Comment (simonbasle) [inline reactor-test/src/main/java/reactor/test/publisher/TestPublisher.java:104]:
```suggestion
	 * subscribers. It buffers the {@link #next(Object)} events and tracks how many
	 * elements have been seen by each subscriber in order to correctly replay the
	 * buffer.
```

**Review by simonbasle — CHANGES_REQUESTED** [review]

File: reactor-test/src/main/java/reactor/test/publisher/ColdTestPublisher.java [hunk 10]
```diff
@@ -226,6 +226,95 @@ void onComplete() {
 			parent.remove(this);
 			actual.onComplete();
 		}
+
+		private void drain() {
+			if (WIP.getAndIncrement(this) == 0) {
+				while (WIP.get(this) > 0) {
```
Comment (simonbasle) [inline reactor-test/src/main/java/reactor/test/publisher/ColdTestPublisher.java:232]: you need to perform that check _atomically_ in the `decrementAndGet` at the end of the loop, rather than here, otherwise it defeats the purpose:
 - thread A wins the guard and goes into the loop
 - thread B looses and exits
 - thread A does a first loop, decrements
 - thread A does a second loop, corresponding to thread B's event
 - **thread A decrements to 0**
 - **thread C enters `drain` and wins the guard (swapping from 0 to 1)**
 - **thread A reads `WIP` and sees `1`**
 - **:bug: both A and C enter the loop and clash**

**Review by simonbasle — CHANGES_REQUESTED** [review]

File: reactor-test/src/main/java/reactor/test/publisher/TestPublisher.java (hunk 6 above)
Comment (simonbasle) [reply reactor-test/src/main/java/reactor/test/publisher/TestPublisher.java]: fixed in `polish: javadoc, formatting and null detection` 84d0afe

**Comment by simonbasle** [comment]
@ericbottard I've made larger improvements to the drain loop and found a couple of bugs along the way

**Review by ericbottard — COMMENTED** [review]
LGTM. The new version of drain() reads much nicer now with the named booleans.

File: reactor-test/src/test/java/reactor/test/publisher/ColdTestPublisherTests.java (hunk 5 above)
Comment (simonbasle) [reply reactor-test/src/test/java/reactor/test/publisher/ColdTestPublisherTests.java:147]: after looking at the test again, it appears it is fine, nevermind

**Review by simonbasle — APPROVED** [review]

**Comment by simonbasle** [comment]
locally tested since we're currently out of credits in Travis CI 😞
